aptos-move/move-examples/nft_auction_house/sources/bid.move

```diff
@@ -0,0 +1,396 @@
+    const ELISTING_EXPIRED: u64 = 3;
+
+    /// Listing hasn't started yet
+    const ELISTING_NOT_STARTED: u64 = 4;
+
+    /// Token amount doesn't match
+    const ETOKEN_AMOUNT_NOT_MATCH: u64 = 5;
+
+    /// Bid doesn't exist
+    const EBID_NOT_EXIST: u64 = 6;
+
+    /// Cannot withdraw fund before bid expiration time
+    const ECANNOT_DRAW_FUND_BEFORE_EXPIRATION_TIME: u64 = 7;
+
+    /// Listing Id doesn't match
+    const ELISTING_ID_NOT_MATCH: u64 = 8;
+
+    /// hold the bid info and coin at user account
+    struct Bid<phantom CoinType> has store {
+        bidder: address,
+        coin: Coin<CoinType>,
+        offer_price: u64,
+        listing_id: ID, // ensure the bid is only executed for the chosen listing_id
+        expiration_sec: u64,
+    }
+
+    /// store all the bids by the user
+    struct BidRecords<phantom CoinType> has key {
+        records: Table<ID, Bid<CoinType>>,
+        bid_event: EventHandle<BidEvent<CoinType>>,
+        withdraw_bid_event: EventHandle<WithdrawBidEvent<CoinType>>,
+    }
+
+    struct BidEvent<phantom CoinType> has copy, drop, store {
+        offer_price: u64,
+        listing_id: ID,
+        bid_id: ID,
+        expiration_sec: u64,
+    }
+
+    struct WithdrawBidEvent<phantom CoinType> has copy, drop, store {
+        bid_id: ID,
+    }
+
+    public fun initialize_bid_records<CoinType>(bidder: &signer)  {
+        let owner_addr = signer::address_of(bidder);
+
+        if (!exists<BidRecords<CoinType>>(owner_addr)) {
+            move_to(
+                bidder,
+                BidRecords<CoinType> {
+                    records: table::new(),
+                    bid_event: account::new_event_handle<BidEvent<CoinType>>(bidder),
+                    withdraw_bid_event: account::new_event_handle<WithdrawBidEvent<CoinType>>(bidder),
+                }
+            );
+        };
+    }
+
+    public fun assert_bid_for_a_listing<CoinType>(
+        token_id: TokenId,
+        offer_price: u64,
+        token_amount: u64,
+        entry: &Listing<CoinType>,
+        bid_time: u64,
+    ) {
+        // validate token_id match
+        assert!(token_id == listing::get_listing_token_id(entry), error::invalid_argument(ETOKEN_ID_NOT_MATCH));
+        // validate offerred amount and price
+        let listed_amount =  listing::get_listing_token_amount(entry);
+        let min_total = listing::get_listing_min_price(entry) * listed_amount;
+        let total_coin_amount = offer_price * token_amount;
+        assert!( total_coin_amount >= min_total, ENO_SUFFICIENT_FUND);
+        assert!(token_amount == listed_amount, ETOKEN_AMOUNT_NOT_MATCH);
+        assert!(bid_time >= listing::get_listing_start(entry), error::invalid_argument(ELISTING_NOT_STARTED));
+        assert!(bid_time <= listing::get_listing_expiration(entry), error::invalid_argument(ELISTING_EXPIRED));
+    }
+
+    /// withdraw the coin and store them in bid struct and return a global unique bid id
+    public fun bid<CoinType>(
```

Comment:
What happens when someone bids multiple times? For example I bid 5 coins and then 15 coins later on. Wouldn't this take 20 coins out of my account? If so, it doesn't seem correct to me as my later higher bids should override.

Reply:
this would just record a bid with a bid_id.  This mainly tries to provide a library for bidding. The actual auction logic can be found in the auction_house example. The auction_house will choose the highest bid to complete the auction